File: javascript/dnd-character/dnd-character.js
export const abilityModifier = (ability) => {
    if (ability < 3) {
        throw new Error('Ability scores must be at least 3');
    }

    if (ability > 18) {
        throw new Error('Ability scores can be at most 18');
    }

    return Math.floor((ability - 10) / 2);
};

export class Character {
    
    static rollAbility() {

        const rolls = Array.from({length: 4}, () => Math.random() * 6);
        return rolls.reduce((a, b) => a + b, 0) - Math.min(...rolls);

    }

    constructor() {
        this._strength = Character.rollAbility();
        this._dexterity = Character.rollAbility();
        this._constitution = Character.rollAbility();
        this._intelligence = Character.rollAbility();
        this._wisdom = Character.rollAbility();
        this._charisma = Character.rollAbility();
        this._hitpoints = 10 + abilityModifier(this._constitution);
    }

    get strength() {
        return this._strength;
    }

    get dexterity() {
        return this._dexterity;
    }

    get constitution() {
        return this._constitution;
    }

    get intelligence() {
        return this._intelligence;
    }

    get wisdom() {
        return this._wisdom;
    }

    get charisma() {
        return this._charisma;
    }

    get hitpoints() {
        return this._hitpoints;
    }
}

File: nim/dnd-character/dnd_character.nim
import std/[math, random, sequtils]

type
  Character* = object
    strength*: int
    dexterity*: int
    constitution*: int
    intelligence*: int
    wisdom*: int
    charisma*: int
    hitpoints*: int

proc ability*: int =
  randomize()
  var rolls = newSeqWith(4, rand(1..6))
  sum(rolls) - min(rolls)
   
proc modifier*(n: int): int =
  floorDiv(n - 10, 2)

proc initCharacter*: Character =
  var ctt = ability()
  Character(
    strength: ability(),
    dexterity: ability(),
    constitution: ctt,
    intelligence: ability(),
    wisdom: ability(),
    charisma: ability(),
    hitpoints: 10 + modifier(ctt)
  )
